### PR TITLE
Feature/backend specific queues v2

### PR DESCRIPTION
**Pull request by diptorupd**
Rebased on top of master.

### CODE REVIEW

**Comment by diptorupd** [comment]
Ready for review.

This PR is the last major thing that I wanted to get into Beta 10. Among other things it makes sure we always know what type of back end we are selecting.

Fixes several important issues:
#53 
#73 
#51 
#33

File: backends/tests/test_sycl_queue_manager.cpp [hunk 1]
```diff
@@ -123,19 +170,33 @@ TEST_F (TestDPPLSyclQueueManager, CheckGetNumActivatedQueues)
...
 
 int
 main (int argc, char** argv)
 {
-  ::testing::InitGoogleTest(&argc, argv);
-  int ret = RUN_ALL_TESTS();
-  return ret;
+    ::testing::InitGoogleTest(&argc, argv);
+    int ret = RUN_ALL_TESTS();
+    return ret;
```
Comment (PokhodenkoSA) [inline backends/tests/test_sycl_queue_manager.cpp:201]: It is possible to link each test case executable with gmain lib and do not create main() function in each cpp file. Lets do it in separate PR.

File: dpctl/tests/test_sycl_kernel_submit.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def test_create_program_from_source (self):
             size_t index = get_global_id(0);                                   \
             c[index] = d*a[index] + b[index];                                  \
         }"
-        with dpctl.device_context(dpctl.device_type.gpu, 0):
+        with dpctl.device_context("opencl:gpu:0"):
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_sycl_kernel_submit.py:40]: Is it possible to set only `"gpu"` ? Is it possible to select default backend?

Comment (PokhodenkoSA) [reply dpctl/tests/test_sycl_kernel_submit.py:40]: I think we need rules for composing this string for developers and users to follow.

**Review by PokhodenkoSA — APPROVED** [review]
Small questions and remarks from me. LGTM.

**Comment by oleksandr-pavlyk** [comment]
Requires gtest 1.10 to build (I was not able to execute `build_for_develop` with gtest 1.8)

**Comment by oleksandr-pavlyk** [comment]
Error handling needs to be improved:

```
In [1]: import dpctl

In [2]: with dpctl.device_context("cpu::0"):
   ...:     print(1)
   ...:
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/localdisk/work/opavlyk/repos/dpctl/dpctl/sycl_core.pyx in dpctl._sycl_core._SyclQueueManager._set_as_current_queue()

KeyError: 'cpu'

During handling of the above exception, another exception occurred:

UnsupportedBackendError                   Traceback (most recent call last)
<ipython-input-2-de76f0cd9073> in <module>
----> 1 with dpctl.device_context("cpu::0"):
      2     print(1)
      3

~/.conda/envs/idp/lib/python3.7/contextlib.py in __enter__(self)
    110         del self.args, self.kwds, self.func
    111         try:
--> 112             return next(self.gen)
    113         except StopIteration:
    114             raise RuntimeError("generator didn't yield") from None

/localdisk/work/repos/dpctl/dpctl/sycl_core.pyx in device_context()

/localdisk/work/repos/dpctl/dpctl/sycl_core.pyx in dpctl._sycl_core._SyclQueueManager._set_as_current_queue()

UnsupportedBackendError: Backend can only be opencl or level-0
```


Note that I am not sure what the arguments of `device_context` changed too, so I was trying something I though might be sensible.

**Comment by oleksandr-pavlyk** [comment]
```
Python 3.7.7 (default, Jul 14 2020, 22:02:37)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.17.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import dpctl

In [2]: with dpctl.device_context('level0:gpu:0'):
   ...:     print(dpctl.get_current_queue().get_sycl_device().dump_device_info())
   ...:
TBB Warning: The number of workers is currently limited to 11. The request for 31 workers is ignored. Further requests for more workers will be silently ignored until the limit changes.

    Name            Intel(R) Gen9
    Driver version  1.0.17767
    Vendor          Intel(R) Corporation
    Profile         FULL_PROFILE
    Device type     gpu
None

In [3]: with dpctl.device_context('level0:gpu:0'):
   ...:     dpctl.get_current_queue().get_sycl_device().dump_device_info()
   ...:
    Name            Intel(R) Gen9
    Driver version  1.0.17767
    Vendor          Intel(R) Corporation
    Profile         FULL_PROFILE
    Device type     gpu

In [4]: with dpctl.device_context('opencl:gpu:0'):
   ...:     dpctl.get_current_queue().get_sycl_device().dump_device_info()
   ...:
    Name            Intel(R) Gen9 HD Graphics NEO
    Driver version  20.35.17767
    Vendor          Intel(R) Corporation
    Profile         FULL_PROFILE
    Device type     gpu

In [5]:
```

Notice the TBB warning on the first invocation. 

It would be nice if `dpctl` provided a function to enumerate all possible arguments understood by `device_context`, even if only to be used to construct an informative error message.

File: backends/include/dppl_sycl_queue_interface.h [hunk 3]
```diff
@@ -36,37 +37,32 @@
...
- * @brief Delete the pointer after casting it to sycl::queue.
+ * @brief Returns the Sycl backend for the provided sycl::queue.
  *
- * @param    QRef           A DPPLSyclQueueRef pointer that gets deleted.
+ * @param    QRef           An opaque pointer to the sycl queue.
+ * @return   A enum DPPLSyclBEType corresponding to the backed for the queue.
  */
 DPPL_API
-void DPPLQueue_Delete (__dppl_take DPPLSyclQueueRef QRef);
+enum DPPLSyclBEType DPPLQueue_GetBackend (__dppl_keep DPPLSyclQueueRef QRef);
```
[the remark this answers is not in the document]
Comment (oleksandr-pavlyk) [reply backends/include/dppl_sycl_queue_interface.h:65]: I was able to remove `enum` there and compile the code just fine.

**Comment by oleksandr-pavlyk** [comment]
I think this is good overall, and should be merged asap.

**Review by oleksandr-pavlyk — APPROVED** [review]
LGTM overall.

File: backends/include/dppl_sycl_queue_interface.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/include/dppl_sycl_queue_interface.h:65]: Right ther are not needed. For some reason, Cython did not like the way I was using typedef-ed enums. So I added them everywhere to be consistent. I can take them out from the C-API.

**Comment by oleksandr-pavlyk** [comment]
The TBB message is coming from `auto ctx = context(d)` line in ``QMgrHelper::init_queues`` when `d` is a GPU device.

```
In [1]: import dpctl

In [2]: with dpctl.device_context('opencl:gpu'):
   ...:     print(1)
   ...:
at the start
before get info
after get info
after get context
before get info
after get info
TBB Warning: The number of workers is currently limited to 11. The request for 31 workers is ignored. Further requests for more workers will be silently ignored until the limit changes.

after get context
before get info
after get info
after get context
before emplace
after emplace
after new
1
```

**Comment by oleksandr-pavlyk** [comment]
Please do not merge. I root caused the TBB messages, and preparing a fix

**Comment by oleksandr-pavlyk** [comment]
Ok, I merge the fix. Please look it over and merge.

File: backends/include/dppl_sycl_queue_interface.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/include/dppl_sycl_queue_interface.h]: Removed the enum in 772b97d

File: backends/include/dppl_sycl_queue_manager.h [hunk 4]
```diff
@@ -63,8 +64,10 @@ __dppl_give DPPLSyclQueueRef DPPLQueueMgr_GetCurrentQueue ();
  * raised if no such device exists.
  */
 DPPL_API
-__dppl_give DPPLSyclQueueRef DPPLQueueMgr_GetQueue (DPPLSyclDeviceType DeviceTy,
-                                                    size_t DNum);
+__dppl_give DPPLSyclQueueRef
+DPPLQueueMgr_GetQueue (enum DPPLSyclBEType BETy,
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/include/dppl_sycl_queue_manager.h]: Changed in 772b97d

File: backends/source/dppl_sycl_context_interface.cpp [hunk 5]
```diff
@@ -36,6 +36,14 @@ namespace
 DEFINE_SIMPLE_CONVERSION_FUNCTIONS(context, DPPLSyclContextRef)
 } /* end of anonymous namespace */
 
+bool DPPLContext_AreEq (__dppl_keep const DPPLSyclContextRef CtxRef1,
+                        __dppl_keep const DPPLSyclContextRef CtxRef2)
+{
+    if(!(CtxRef1 && CtxRef2))
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_context_interface.cpp:42]: If either context ref is NULL then it indicates some corrupted state somewhere. So, NULL == NULL is a tricky case. Mostly we should return false in that case.

But, let us tackle this in #35.

File: backends/source/dppl_sycl_context_interface.cpp [hunk 6]
```diff
@@ -46,3 +54,23 @@ void DPPLContext_Delete (__dppl_take DPPLSyclContextRef CtxRef)
 {
     delete unwrap(CtxRef);
 }
+
+DPPLSyclBEType
+DPPLContext_GetBackend (__dppl_keep const DPPLSyclContextRef CtxRef)
+{
+    auto BE = unwrap(CtxRef)->get_platform().get_backend();
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_context_interface.cpp:61]: Agreed. Let us tackle in #35

File: backends/source/dppl_sycl_platform_interface.cpp [hunk 7]
```diff
@@ -27,95 +27,162 @@
...
+    auto i = 0ul;
+    for (auto be : be_set) {
+        BEArr[i] = be;
+        ++i;
+    }
+
+    return BEArr;
+}
+
+void DPPLPlatform_DeleteListOfBackends (__dppl_take enum DPPLSyclBEType *BEArr)
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_platform_interface.cpp]: The problem I was worried about is how do we delete the array from Cython? Does Cython provide `delete[]`?

File: backends/tests/test_sycl_queue_manager.cpp (hunk 1 above)
Comment (diptorupd) [reply backends/tests/test_sycl_queue_manager.cpp:201]: Opened a new issues #99

File: dpctl/tests/test_sycl_kernel_submit.py (hunk 2 above)
Comment (diptorupd) [reply dpctl/tests/test_sycl_kernel_submit.py:40]: I did not add the option for default backend yet. Let us do as a follow up. Opened #100

File: backends/source/dppl_sycl_device_interface.cpp [hunk 8]
```diff
@@ -56,20 +56,26 @@ void dump_device_info (const device & Device)
...
+    ss << "accelerator" << '\n';
+    break;
+    case info::device_type::custom:
+    ss << "custom" << '\n';
+    break;
+    case info::device_type::host:
+    ss << "host" << '\n';
+    break;
+    default:
+    ss << "unknown" << '\n';
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_device_interface.cpp]: Fixed in 81dd13d

File: backends/source/dppl_sycl_platform_interface.cpp [hunk 9]
```diff
@@ -27,95 +27,162 @@
...
 
+namespace
+{
+std::set<DPPLSyclBEType>
+get_set_of_backends ()
+{
+    std::set<DPPLSyclBEType> be_set;
+    for (auto p : platform::get_platforms()) {
+		if(p.is_host())
+			continue;
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_platform_interface.cpp]: Fixed now?

File: backends/source/dppl_sycl_platform_interface.cpp [hunk 10]
```diff
@@ -27,95 +27,162 @@
...
+            break;
+            case info::device_type::custom:
+            ss << "custom" << '\n';
+            break;
+            case info::device_type::host:
+            ss << "host" << '\n';
+            break;
+            default:
+            ss << "unknown" << '\n';
+            }
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_platform_interface.cpp]: Fixed the formattinh in 81dd13d

[the remark this answers is not in the document]
Comment (diptorupd) [reply backends/source/dppl_sycl_platform_interface.cpp]: I am opening a new issue #102  to track the refactoring.

File: backends/source/dppl_sycl_platform_interface.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (PokhodenkoSA) [reply backends/source/dppl_sycl_platform_interface.cpp]: Agree. I have googled and did not find another way.

[the remark this answers is not in the document]
Comment (oleksandr-pavlyk) [reply backends/source/dppl_sycl_platform_interface.cpp]: Cython does not provide `delete[]`.

**Comment by diptorupd** [comment]
Merging now. Thanks for the reviews!
